### PR TITLE
Migrate safe models

### PR DESCRIPTION
**Pull request by Hosuke**
Migrate all spells under models/safe/*
except `safe_ethereum_airdrop_claims`.

### CODE REVIEW

**Comment by Hosuke** [comment]
TODO: partition_by block_month

**Comment by jeff-dude** [comment]
@danielpartida wanted to ensure you were aware of this PR migrating safe spells -- any chance you can help review?

**Comment by danielpartida** [comment]
Thank you for the heads up @jeff-dude. Highly appreciated it. I will review this PR in detail over weekend and share my feedback

**Comment by danielpartida** [comment]
Hi @jeff-dude and @Hosuke 

I added my comments for the Safe Arbitrum spells but the same applies to all the other chains. Please let me know whether these comments make sense to you.

And thank you for the help on migrating these spells to DuneSQL

**Comment by Hosuke** [comment]
> Hi @jeff-dude and @Hosuke
> 
> I added my comments for the Safe Arbitrum spells but the same applies to all the other chains. Please let me know whether these comments make sense to you.
> 
> And thank you for the help on migrating these spells to DuneSQL

Thank you for your reviews. But currently I cannot find those comments in the PR.
Sorry but can you comment again? I will follow the suggestions you make.

**Comment by jeff-dude** [comment]
> Hi @jeff-dude and @Hosuke
> 
> I added my comments for the Safe Arbitrum spells but the same applies to all the other chains. Please let me know whether these comments make sense to you.
> 
> And thank you for the help on migrating these spells to DuneSQL

the comments are likely still in pending state, please do finalize and submit review and we'll be happy to help continue!

**Comment by danielpartida** [comment]
Sorry about that, please let me know if you are able to see the comments now @jeff-dude @Hosuke

File: models/safe/arbitrum/safe_arbitrum_eth_transfers.sql [hunk 1]
```diff
@@ -1,16 +1,17 @@
 {{ 
     config(
         materialized='incremental',
+        tags = ['dunesql'],
         alias = alias('eth_transfers'),
-        partition_by = ['block_date'],
+        partition_by = ['block_month'],
```
[the remark this answers is not in the document]
Comment (jeff-dude) [reply models/safe/arbitrum/safe_arbitrum_eth_transfers.sql:6]: we are looking to universally increase our partition size and have landed on `block_month` as a consistent approach. we're either removing completely (if smaller dataset) or increasing to monthly level at least

File: models/safe/arbitrum/safe_arbitrum_safes.sql [hunk 2]
```diff
@@ -1,51 +1,53 @@
...
 from {{ source('arbitrum', 'traces') }} et 
 join {{ ref('safe_arbitrum_singletons') }} s
     on et.to = s.address
 where et.success = true
     and et.call_type = 'delegatecall' -- delegatecall to singleton is Safe (proxy) address
-    and substring(et.input, 0, 10) in (
-        '0x0ec78d9e', -- setup method v0.1.0
-        '0xa97ab18a', -- setup method v1.0.0
-        '0xb63e800d' -- setup method v1.1.0, v1.1.1, v1.2.0, v1.3.0, v1.3.0L2
+    and bytearray_substring(et.input, 1, 4) in (
```
[the remark this answers is not in the document]
Comment (Hosuke) [reply models/safe/arbitrum/safe_arbitrum_safes.sql:42]: I will provide a query proof for this one.
It is a bit tricky.

File: models/safe/arbitrum/safe_arbitrum_eth_transfers.sql [hunk 3]
```diff
@@ -19,43 +20,45 @@
 select 
     s.address,
     try_cast(date_trunc('day', et.block_time) as date) as block_date,
+    CAST(date_trunc('month', et.block_time) as DATE) as block_month,
     et.block_time,
-    -et.value as amount_raw,
+    -CAST(et.value AS INT256) as amount_raw,
```
[the remark this answers is not in the document]
Comment (jeff-dude) [reply models/safe/arbitrum/safe_arbitrum_eth_transfers.sql:25]: in the traces table, `value` is `uint256` -- if we need to allow negative, then i believe `int256` is the right choice here

File: models/safe/arbitrum/safe_arbitrum_safes.sql [hunk 4]
```diff
@@ -1,51 +1,53 @@
...
+        when et.to = 0x34cfac646f301356faa8b21e94227e3583fe3f5f then '1.1.1'
+        when et.to = 0x6851d6fdfafd08c0295c392436245e5bc78b0185 then '1.2.0'
+        when et.to = 0xd9db270c1b5e3bd161e8c8503c55ceabee709552 then '1.3.0'
+        when et.to = 0x69f4d1788e39c87893c980c06edf4b7f686e2938 then '1.3.0'  -- for chains with EIP-155
+        when et.to = 0x3e5c63644e683549055b9be8653de26e0b4cd36e then '1.3.0L2'
+        when et.to = 0xfb1bffc9d739b8d520daf37df666da4c687191ea then '1.3.0L2' -- for chains with EIP-155
         else 'unknown'
     end as creation_version,
     try_cast(date_trunc('day', et.block_time) as date) as block_date,
+    CAST(date_trunc('month', et.block_time) as DATE) as block_month,
```
[the remark this answers is not in the document]
Comment (jeff-dude) [reply models/safe/arbitrum/safe_arbitrum_safes.sql:34]: as noted above, it's a new approach we are introducing to increase partition sizes

File: models/safe/arbitrum/safe_arbitrum_safes.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeff-dude) [reply models/safe/arbitrum/safe_arbitrum_safes.sql:42]: `input` is varbinary, so `bytearray_substring()` is the right function here

from what i've seen, the translation from `substring` to `bytearray_substring` usually is:
first number: subtract 1, divide by 2
second number: divide by 2
here is an example i'm working through:
https://github.com/duneanalytics/spellbook/pull/3984/files#diff-51fae7300141bbf294d04a507cc9c81b3e72bca1e523a912c3343aaa43bfcd64R21

@Hosuke can you confirm here?

File: models/safe/arbitrum/safe_arbitrum_schema.yml [hunk 5]
```diff
@@ -61,6 +63,7 @@ models:
     columns:
       - *blockchain
       - *block_date
+      - *block_month
```
[the remark this answers is not in the document]
Comment (jeff-dude) [reply models/safe/arbitrum/safe_arbitrum_schema.yml:66]: similar answer here

File: models/safe/arbitrum/safe_arbitrum_safes.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hosuke) [reply models/safe/arbitrum/safe_arbitrum_safes.sql:42]: I can confirm here.
This is the test I did during the migration:
https://dune.com/queries/2823326

[the remark this answers is not in the document]
Comment (Hosuke) [reply models/safe/arbitrum/safe_arbitrum_safes.sql:42]: If the first number is `0`, we need to make them `1`, as for `0x...` place.

If the first number is not `0`, we can convert the parameters //2.

[the remark this answers is not in the document]
Comment (jeff-dude) [reply models/safe/arbitrum/safe_arbitrum_safes.sql:42]: yep, makes sense to me!

[the remark this answers is not in the document]
Comment (danielpartida) [reply models/safe/arbitrum/safe_arbitrum_safes.sql:42]: Ok, they are equivalent https://dune.com/queries/2857157. Looks good to me

**Comment by danielpartida** [comment]
Thank you for the feedback and responses @jeff-dude @Hosuke. All looks good to me.

One last question since we are migrating/updating these Safe spells. For this [Safe Celo transfer spell](https://spellbook-docs.dune.com/#!/model/model.spellbook.safe_celo_celo_transfers#details), could we change the alias from "transfers" to "celo_transfers". It would then have the same structure as the other Safe transfer spells, e.g. this [Safe ETH transfer spell](https://spellbook-docs.dune.com/#!/model/model.spellbook.safe_ethereum_eth_transfers)

Screenshot 2023-08-15 at 09 19 21

**Comment by Hosuke** [comment]
> Safe ETH transfer spell

I see, so I will make the table available as `safe_celo.celo_transfers` in dune sql engine.

File: models/safe/safe_safes_all.sql [hunk 6]
```diff
@@ -11,6 +12,7 @@
  ref('safe_arbitrum_safes')
 ,ref('safe_avalanche_c_safes')
 ,ref('safe_bnb_safes')
+,ref('safe_celo_safes')
```
Comment (Hosuke) [inline models/safe/safe_safes_all.sql:15]: Add `safe_celo_safes` here as well.

**Comment by Hosuke** [comment]
Sample safe_celo.celo_transfers:
https://dune.com/queries/2857956

**Review by jeff-dude — APPROVED** [review]
thanks for review @danielpartida
& thanks for the migration @Hosuke -- great stuff!

**Pull request merged**